### ISSUE TITLE
Quadlet: allow specifying `--force` for the `podman kube down` command in podman-systemd `.kube` file

### ISSUE THREAD

**Issue by zephyros-dev** [containers/podman#20284]
### Feature request description

Currently there's no option for changing the arguments for the `podman kube down` in the generated systemd file for tearing down the service

### Suggest potential solution

Currently the only argument available is [--force](https://docs.podman.io/en/latest/markdown/podman-kube-down.1.html), so there are 2 options I can think of:
- Add --force as an named argument in the .kube file (Force or DownForce?)
- Allow specifying extra arguments like the `PodmanArgs` for the podman down command (PodmanDownArgs)

### Additional context

I can try to implement this feature, just need a discussion on which solution to take

**Comment by vrothberg** [comment]
Thanks for reaching out, @zephyros-dev !

I am open to allow for tweaking the `kube down` command.  I guess a `KubeDownForce=true` field would do the trick as well.

@ygalblum WDYT?

**Comment by rhatdan** [comment]
PodmanArgs should also be available in .kube files if it is not now.

**Comment by rhatdan** [comment]
Looks like it is allowed.

**Comment by zephyros-dev** [comment]
Regarding the default behavior, here's how I pictured it would be like, coming from docker-compose:
1. Running `docker compose up` will start the container and create the volume defined in the file
- This is akin to running `systemctl start` with the PersistentVolumeClaim defined in the Kube yaml file
3. Running `docker compose down`  will stop the container but does NOT delete the volume
- This is akin to setting `KubeDownForce` to false and running `systemctl stop`
4. Running `docker compose down -v` will stop the container and delete the volume
- This is akin to setting `KubeDownForce` to true and running `systemctl stop`
5. Running `docker compose restart` will restart all the container but does NOT recreate the volume
- The volume will not be recreated when running `systemctl restart` regardless of the `KubeDownForce` settings
6. Container that are run on system startup does not recreate the volume
During my testing, I haven't found a way to replicate the 3, 4 and 5 behavior using systemd. The current quadlet implementation setup the `kube down` command in `ExecStop`, which is ran during both `restart` and `stop` command. With `KubeDownForce=true`, the volume will be deleted whenever the service is restarted, which is a surprising behavior for user coming from docker-compose (at least it is for me).
To replicate the docker-compose behavior, I think setting the `--force` command to be run ONLY on `systemctl stop` (and not on restart or startup) would do the trick. AFAIK I haven't found a way to do so.
Going forward, if the above implementation is not possible, personally I don't think the `KubeDownForce` setting is useful enough to be included in the file, and could even introduced surprising behavior. As an alternative for tearing down the volume defined in the setup file, one could run the `podman kube down --force` manually on the kubernetes manifest (which is akin to running `docker-compose down -v`)

**Comment by ygalblum** [comment]
AFAIK, systemd does not support `restart` specific configuration and the `restart` command is translated into a sequence of `stop`->`start`. As a result, there is no way to define different behaviors when running in a single systemd unit. **Please correct me if I'm wrong here**

Having said that, I have an idea that can add support for this feature by splitting the unit into two Quadlets, one for the volume (with `KubeDownForce=true`) and one for the application (without setting `KubeDownForce`), setting the application one to depend on the volume one.

This way calling `systemctl start` on the application's service will also start the volume service. Stopping the application's service will not stop that of the volume and as a result will not remove the volume and the same for when it is restarted. In order to remove the volume the user will need to explicitly stop the volume service.

In order to support this idea, there are two changes to make:
1. Add support for the `KubeDownForce` key
2. Currently, `.kube` file based services fail when no containers are defined in the YAML file because Quadlet sets the Service Type to `notify`. Quadlet already supports manual setting of the `Type` field for `.container` files and the same logic needs to be applied for `.kube` file.

WDYT?

**Comment by vrothberg** [comment]
I'd prefer exploring whether it's possible to query systemd whether the unit is being restarted or not.  Podman could do such a check when running inside of a unit.  That sounds less complex to me.  A `restart` is similar to `stop+start` but certain resources are not cleaned up, so I hope we could distinguish the two.

**Comment by ygalblum** [comment]
I see your point. But, wouldn't that open a Pandora box? I mean this requires Podman not only to know that the `down` operation is a part of a restart but also decide which flags are meant only for stop and not for restart.

**Comment by vrothberg** [comment]
I think this can be done in `cmd/podman/kube` (but only for local and not for remote client).  I am not married to this idea at all but I think we should explore more alternatives.

Creating a separate `.volume` unit sounds a bit too complex.  Maybe we'll go with it but I want to make sure we won't miss a simpler alternative.

**Comment by zephyros-dev** [comment]
@ygalblum Personally I think there are a few issues with the `.volume` file implementation:
- In term of UX, the definition of the the volume in the case of `.kube` file is already defined in the kubernetes pvc object, and I think most users would rather managing it from there rather than having to keep track of another volume service.
- Currently the way I see it is that the quadlet files are made to declare the resources like kuberenetes manifest. Correct me if I misunderstood, but the `.volume` implementation would automate the creation and removal of the quadlet files from the definition of the pvc object, which make the declarative properties no longer true and confuses the user (where did this `.volume` service came from, manually or a `.kube` service?)
- Correct me if I'm wrong, but I believe each `.volume` service can only handle a single named volume. A single `.kube` service can have multiple pvc object, which will creates multiple volumes, so creating/stopping it would have to handle multiple `.volume` services and quadlet file
All in all, I think most user would prefer a method that only deal with one service for the life cycle of everything, so I'd prefer @vrothberg approach. Granted It will probably be a complex solution, but I think a complex implementation with simple interface is preferable to the inverse.

**Comment by ygalblum** [comment]
@zephyros-dev I agree about UX parts of your comment. Obviously handing a single service is preferable. But, I was focusing on what is available since I do not know of a way to differentiate between restart and stop. According to @vrothberg there are ways to do that, and if so, that's great.

Having said that, I do want to emphasis that I did not talk about `.volume` units, but instead of **one** additional `.kube` unit file. My suggestion is to take all PVCs (one or more) from the original YAML file and place them under a new K8S YAML file. Then have two `.kube` units, one for the volume(s) and the other for the application.

As I wrote, currently, `.kube` based units generate an error if the YAML file does not contain any containers (due to the need to change the Type to `oneshot`). Now, since `KubeDownForce` is not supported yet either, the current version does not fulfill the requirement anyhow. However, these fixes are fairly simple and we could support the less-UX-friendly option with a minimal effort.
